File: grokkingfp-in-clojure/src/lagenorhynque/ch04_word_scoring.clj
(ns lagenorhynque.ch04-word-scoring
  "第4章 値としての関数"
  (:require
   [clojure.spec.alpha :as s]))

;; 4.22 コーヒーブレイク：Scalaで関数を渡す

(s/fdef ranked-words
  :args (s/cat :word-score-fn (s/fspec :args (s/cat :word string?)
                                       :ret number?)
               :words (s/coll-of string?))
  :ret (s/coll-of string?))

(defn ranked-words [word-score-fn words]
  (->> words
       (sort-by word-score-fn)
       reverse))

(s/fdef score
  :args (s/cat :word string?)
  :ret number?)

(defn score [word]
  (->> word
       (remove #{\a})
       count))

(s/fdef bonus
  :args (s/cat :word string?)
  :ret number?)

(defn bonus [word]
  (if (some #{\c} word)
    5
    0))

(s/fdef penalty
  :args (s/cat :word string?)
  :ret number?)

(defn penalty [word]
  (if (some #{\s} word)
    7
    0))

File: grokkingfp-in-haskell/src/Ch04_WordScoring.hs
{- |
第4章 値としての関数
-}
module Ch04_WordScoring(module Ch04_WordScoring) where

import           Data.List (sortOn)

-- 4.22 コーヒーブレイク：Scalaで関数を渡す

rankedWords :: (String -> Int) -> [String] -> [String]
rankedWords scoreFn = reverse . sortOn scoreFn

score :: String -> Int
score = length . filter (/= 'a')

bonus :: String -> Int
bonus word = if 'c' `elem` word then 5 else 0

penalty :: String -> Int
penalty word = if 's' `elem` word then 7 else 0
